### PR TITLE
refactor(bigtable)!: modernize Instance Admin

### LINKED ISSUE #7534: Consider implementing `InstanceAdmin`, `InstanceAdminClient` in terms of generated `BigtableInstanceAdminConnection`
Basically the same as #7530, but without needing to deal with an async stub call

### PR DESCRIPTION
**Pull request by dbolduc**
Fixes #7534.

This PR is a doozy. Apologies in advance to the reviewer. Changing the connection layer meant that the unit test for the `InstanceAdmin` calls had to be written from scratch.

Implement `InstanceAdmin` on top of the generated Connection class. Reduce `InstanceAdminClient` to be a Connection + project id. This gives us all the features of the generated libraries (like GUAC. yay).

For some reason, `check-api-pr` worked for me locally? I feel like it shouldn't have though.

A minor detail to note:
`internal/async_retry_unary_rpc_and_poll.h` was used by both `InstanceAdmin` and `TableAdmin`. Now it is only used by `TableAdmin`.

The test (`internal/async_retry_unary_rpc_and_poll_test.cc`) was written in terms of `InstanceAdmin`. I decided to delete it entirely, even though the class is still in use, because it is not worth temporarily rewriting. The class will be deleted upon conversion of `TableAdmin`.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8097)

### CODE REVIEW

File: google/cloud/bigtable/instance_admin_client.h [hunk 1]
```diff
@@ -18,321 +18,39 @@
...
+ * maintained only for backwards compatibility.
  *
- * @par Cost
- * Applications should avoid unnecessarily creating new objects of type
- * `InstanceAdminClient`. Creating a new object of this type typically requires
- * connecting to the Cloud Bigtable servers, and performing the authentication
- * workflows with Google Cloud Platform. These operations can take many
- * milliseconds, therefore applications should try to reuse the same
- * `InstanceAdminClient` instances when possible.
+ * @deprecated Please consider using
```
Comment (coryan) [inline google/cloud/bigtable/instance_admin_client.h:38]: In a future PR, maybe we should add a `[[deprecated]]` attribute (or its equivalent for compilers that do not support that).

File: CHANGELOG.md [hunk 2]
```diff
@@ -90,10 +90,48 @@
...
+If more than your tests are broken, please use
+`bigtable_admin::BigtableInstanceAdminClient` in favor of
+`bigtable::InstanceAdmin`, and `bigtable_admin::BigtableInstanceAdminConnection`
+in favor of `bigtable::InstanceAdminClient`. These classes will incorporate the
+newest features of both the [Cloud Bigtable Admin API] and the C++ client
+library. For more information on these new classes, see our
+[Architecture Design] document.
+
+Again, we apologize for making this breaking change, but we believe it is in the
+best interest of our customers.
```
Comment (coryan) [inline CHANGELOG.md:126]: s/best interest/best long-term interests/ ?

File: CHANGELOG.md [hunk 3]
```diff
@@ -90,10 +90,48 @@
...
+**BREAKING CHANGE:** The `bigtable::InstanceAdminClient` interface has changed
+significantly. Any code that extends this class or calls its experimental public
+APIs (`reset()`, `Channel()`) will be broken. For the most part, this should
+only affect customers who mock this class in their tests. Code that calls
+`bigtable::MakeInstanceAdminClient()` or
+`bigtable::CreateDefaultInstanceAdminClient()` will continue to work as before.
+
+This change will allow us to deliver new features more quickly by reducing the
+maintenance costs of the library. It also provides a better mocking interface
+for `bigtable::InstanceAdmin`. We are sorry to those that are inconvenienced by
```
Comment (coryan) [inline CHANGELOG.md:104]: s/We are sorry to those that are inconvenienced/We apologize if you are inconvenienced/

File: google/cloud/bigtable/instance_admin.cc [hunk 4]
```diff
@@ -33,209 +25,82 @@ namespace google {
...
-  } while (!page_token.empty());
-
-  if (!status.ok()) {
-    return MakeStatusFromRpcError(status);
-  }
-
-  std::move(unique_failed_locations.begin(), unique_failed_locations.end(),
+  btadmin::ListInstancesRequest request;
+  request.set_parent(project_name());
+  auto sor = connection_->ListInstances(request);
```
Comment (coryan) [inline google/cloud/bigtable/instance_admin.cc:39]: The previous code iterates over the results, possibly making multiple calls. I do not think `Connection` does that?

**Review by coryan — CHANGES_REQUESTED** [review]

**Review by coryan — CHANGES_REQUESTED** [review]
I am continuing the review in Reviewable given the complexity of the change.

> Reviewed 19 of 19 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8097#-:-MuHqL0b7RXWMzW52GMT:bh9vjz0)* status: all files reviewed, 4 unresolved discussions (waiting on @dbolduc)

___
*[google/cloud/bigtable/instance_admin.cc, line 39 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8097#-MuHejLnrk8dqdAto2TE-r1-39:-MuHpNL9CsNYivfc3zB-:b-l4tade) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4120cbe205b21d58b6d8210b6557355e87d36fe6/google/cloud/bigtable/instance_admin.cc#L39)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

The previous code iterates over the results, possibly making multiple calls. I do not think `Connection` does that?
</blockquote></details>

FWIW, the iteration behavior was intentional, but also, it is probably hard to catch because in most cases the first request contains all responses.
___
*[google/cloud/bigtable/instance_admin_client.h, line 33 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8097#-MuHpkK22mIBbQpxXwIy:-MuHpkK22mIBbQpxXwIz:bubdizs) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4120cbe205b21d58b6d8210b6557355e87d36fe6/google/cloud/bigtable/instance_admin_client.h#L33)):*
> ```objc
>  *
>  * This class is used as a conduit to pass configuration options to
>  * `bigtable::InstanceAdmin`. Formerly, this class was responsible for
> ```

I would skip what this class used to do, just say that it is deprecated and what we recommend instead.

File: google/cloud/bigtable/instance_admin.cc (hunk 4 above)
Comment (dbolduc) [reply google/cloud/bigtable/instance_admin.cc:39]: (responding only to this comment) `ListInstances` is no longer a paginated api. See: https://github.com/googleapis/googleapis/blob/853e8a29cecd0816c8b003b299032c3a5168ef8b/google/bigtable/admin/v2/bigtable_instance_admin.proto#L347-L348

This is also true of `ListClusters`: https://github.com/googleapis/googleapis/blob/853e8a29cecd0816c8b003b299032c3a5168ef8b/google/bigtable/admin/v2/bigtable_instance_admin.proto#L436-L437

File: google/cloud/bigtable/instance_admin_client.h (hunk 1 above)
Comment (dbolduc) [reply google/cloud/bigtable/instance_admin_client.h:38]: Ack.

**Review by dbolduc — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8097)* status: 16 of 19 files reviewed, 4 unresolved discussions (waiting on @coryan)

___
*[CHANGELOG.md, line 104 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8097#-MuHejLnrk8dqdAto2T9-r1-104:-MuI2ptB5bXoN_B-bALS:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4120cbe205b21d58b6d8210b6557355e87d36fe6/CHANGELOG.md#L104)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

s/We are sorry to those that are inconvenienced/We apologize if you are inconvenienced/
</blockquote></details>

Done.
___
*[CHANGELOG.md, line 126 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8097#-MuHejLnrk8dqdAto2T9-r1-126:-MuI2qXE2jVtWYXNUrYz:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4120cbe205b21d58b6d8210b6557355e87d36fe6/CHANGELOG.md#L126)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

s/best interest/best long-term interests/ ?
</blockquote></details>

Done.
___
*[google/cloud/bigtable/instance_admin_client.h, line 33 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8097#-MuHpkK22mIBbQpxXwIy:-MuI3niT-n4RXHKi9f43:b-rlpgik) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4120cbe205b21d58b6d8210b6557355e87d36fe6/google/cloud/bigtable/instance_admin_client.h#L33)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I would skip what this class used to do, just say that it is deprecated and what we recommend instead.
</blockquote></details>

Done.

And whoops, the documentation was stale. I updated it to reflect the current design.

**Review by coryan — APPROVED** [review]
:lgtm:

> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8097#-:-MuICBQo6sbliWyS4XR0:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @dbolduc)

___
*[google/cloud/bigtable/instance_admin.cc, line 39 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8097#-MuHejLnrk8dqdAto2TE-r1-39:-MuIBauoAMh9CdjMzEU5:b-rmw1b3) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4120cbe205b21d58b6d8210b6557355e87d36fe6/google/cloud/bigtable/instance_admin.cc#L39)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

(responding only to this comment) `ListInstances` is no longer a paginated api. See: https://github.com/googleapis/googleapis/blob/853e8a29cecd0816c8b003b299032c3a5168ef8b/google/bigtable/admin/v2/bigtable_instance_admin.proto#L347-L348

This is also true of `ListClusters`: https://github.com/googleapis/googleapis/blob/853e8a29cecd0816c8b003b299032c3a5168ef8b/google/bigtable/admin/v2/bigtable_instance_admin.proto#L436-L437
</blockquote></details>

TIL. Thanks.
___
*[google/cloud/bigtable/instance_admin_test.cc, line 61 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8097#-MuIAzxL3Wj06eXak6ej:-MuIAzxL3Wj06eXak6ek:b-wjzsif) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/60ffeb05df0f2cd5493de7593797ad79da8fd265/google/cloud/bigtable/instance_admin_test.cc#L61)):*
> ```cpp
> 
> std::string const kProjectId = "the-project";
> std::string const kInstanceId = "the-instance";
> ```

For a future cleanup: all these globals with non-trivial destructors are a violation of the Google Style Guide.  I happen to think that the style guide is wrong in this respect, but you may want to consider cleaning it up.  Alternatively, they are fine as part of the test fixture :shrug:

**Comment by dbolduc** [comment]
> For a future cleanup: all these globals with non-trivial destructors are a violation of the Google Style Guide. I happen to think that the style guide is wrong in this respect, but you may want to consider cleaning it up. Alternatively, they are fine as part of the test fixture

Ugh, I should remember having done the original "cleanup". Also seems like I missed a bunch of `(auto|std::string) const k...` in anonymous namespaces.... I will get to them in 2023.

**Review by coryan — APPROVED** [review]
Still :lgtm: 

> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8097#-:-MuIaxZiDCe7vp0Ahg-O:b-dbkjmx)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @dbolduc)

**Pull request merged**